src/vs/editor/contrib/linesOperations/common/linesOperations.ts

```diff
@@ -401,3 +407,267 @@ export class DeleteAllLeftAction extends EditorAction {
 		editor.executeEdits(this.id, edits);
 	}
 }
+
+@editorAction
+export class JoinLinesAction extends EditorAction {
+	constructor() {
+		super({
+			id: 'editor.action.joinLines',
+			label: nls.localize('lines.joinLines', "Join Lines"),
+			alias: 'Join Lines',
+			precondition: EditorContextKeys.Writable,
+			kbOpts: {
+				kbExpr: EditorContextKeys.TextFocus,
+				primary: KeyMod.WinCtrl | KeyCode.KEY_J
+			}
+		});
+	}
+
+	public run(accessor: ServicesAccessor, editor: ICommonCodeEditor): void {
+		let selection = editor.getSelection();
+		let model = editor.getModel();
+		let startLineNumber = selection.startLineNumber;
+		let startColumn = 1;
+		let endLineNumber: number,
+			endColumn: number,
+			columnDeltaOffset: number;
+
+		let selectionEndPositionOffset = model.getLineContent(selection.endLineNumber).length - selection.endColumn;
+
+		if (selection.isEmpty() || selection.startLineNumber === selection.endLineNumber) {
+			let position = selection.getStartPosition();
+			if (position.lineNumber < model.getLineCount()) {
+				endLineNumber = startLineNumber + 1;
+				endColumn = model.getLineMaxColumn(endLineNumber);
+			} else {
+				endLineNumber = position.lineNumber;
+				endColumn = model.getLineMaxColumn(position.lineNumber);
+			}
+		} else {
+			endLineNumber = selection.endLineNumber;
+			endColumn = model.getLineMaxColumn(endLineNumber);
+		}
+
+		let trimmedLinesContent = model.getLineContent(startLineNumber);
+
+		for (let i = startLineNumber + 1; i <= endLineNumber; i++) {
+			let lineText = model.getLineContent(i);
+			let firstNonWhitespaceIdx = model.getLineFirstNonWhitespaceColumn(i);
+
+			if (firstNonWhitespaceIdx >= 1) {
+				let insertSpace = true;
+				if (trimmedLinesContent === '') {
+					insertSpace = false;
+				}
+
+				if (trimmedLinesContent.charAt(trimmedLinesContent.length - 1) === ' ' ||
+					trimmedLinesContent.charAt(trimmedLinesContent.length - 1) === '\t') {
+					insertSpace = false;
+					trimmedLinesContent = trimmedLinesContent.replace(/[\s\uFEFF\xA0]+$/g, ' ');
+				}
+
+				let lineTextWithoutIndent = lineText.substr(firstNonWhitespaceIdx - 1);
+
+				if (lineTextWithoutIndent.charAt(0) === ')') {
```

Comment:
Why are we looking for `)` in the line text here ? i.e. this looks weird to me, why not for `]`. I suggest to remove this unless it is part of some weird join lines spec.